File: prototype/php/src/app/Actions/Cart/RemoveFromCartTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Cart;

it('takes a listing back out of the cart', function (): void {
    $cart = $this->cartFor($this->verifiedCustomer());
    $seller = $this->seller();
    $kept = $this->listing($seller);
    $removed = $this->listing($seller);
    $addToCart = app(AddToCart::class);
    $now = $this->moment('2026-08-20 09:00:00');
    $addToCart($cart, $kept, 1, $now);
    $addToCart($cart, $removed, 1, $now);

    app(RemoveFromCart::class)($cart, $removed);

    expect($cart->items()->pluck('listing_id')->all())->toBe([$kept->id]);
});

it('changes nothing when removing a listing the cart never held', function (): void {
    $cart = $this->cartFor($this->verifiedCustomer());
    $listing = $this->listing($this->seller());

    app(RemoveFromCart::class)($cart, $listing);

    expect($cart->items()->count())->toBe(0);
});

File: prototype/rails/src/test/actions/carts/remove_from_cart_test.rb
require "test_helper"

module Carts
  class RemoveFromCartTest < ActiveSupport::TestCase
    test "it takes the listing out of the cart" do
      shop = create_seller
      kept = create_listing(shop)
      dropped = create_listing(shop)
      cart = cart_holding(create_verified_customer, kept, dropped)

      RemoveFromCart.new.call(cart: cart, listing: dropped)

      assert_equal [kept], cart.reload.items.map(&:listing)
    end

    test "removing a listing the cart never held changes nothing" do
      cart = cart_holding(create_verified_customer, create_listing)

      RemoveFromCart.new.call(cart: cart, listing: create_listing)

      assert_equal 1, cart.reload.items.count
    end
  end
end
